cmd/agent/main.go

```diff
@@ -1,3 +1,220 @@
+	},
+	"MSpanInuse": &metric{
+		metricType: "gauge",
+	},
+	"MSpanSys": &metric{
+		metricType: "gauge",
+	},
+	"Mallocs": &metric{
+		metricType: "gauge",
+	},
+	"NextGC": &metric{
+		metricType: "gauge",
+	},
+	"NumForcedGC": &metric{
+		metricType: "gauge",
+	},
+	"NumGC": &metric{
+		metricType: "gauge",
+	},
+	"OtherSys": &metric{
+		metricType: "gauge",
+	},
+	"PauseTotalNs": &metric{
+		metricType: "gauge",
+	},
+	"StackInuse": &metric{
+		metricType: "gauge",
+	},
+	"StackSys": &metric{
+		metricType: "gauge",
+	},
+	"Sys": &metric{
+		metricType: "gauge",
+	},
+	"TotalAlloc": &metric{
+		metricType: "gauge",
+	},
+	"PollCount": &metric{
+		metricType: "counter",
+		value:      "0",
+	},
+	"RandomValue": &metric{
+		metricType: "gauge",
+	},
+}
+
+func updateMetrics(metrics map[string]*metric, statStruct *runtime.MemStats, mutex *sync.Mutex) error {
+	mutex.Lock()
+	runtime.ReadMemStats(statStruct)
+	statStructFields := reflect.ValueOf(statStruct).Elem()
+
+	for i := 0; i < statStructFields.NumField(); i++ {
+
+		field := statStructFields.Field(i)
+		fieldName := statStructFields.Type().Field(i).Name
+		fieldType := field.Kind()
+		if metric, ok := metrics[fieldName]; ok {
+			switch fieldType {
+			case reflect.Float64:
+				metric.value = strconv.FormatFloat(field.Float(), 'f', -1, 64)
+			case reflect.Uint64:
+				metric.value = strconv.FormatUint(field.Uint(), 10)
+			case reflect.Uint32:
+				metric.value = strconv.FormatUint(field.Uint(), 10)
+			}
+		}
+
+	}
+
+	pollCountValue, err := strconv.ParseInt(metrics["PollCount"].value, 10, 64)
+	if err != nil {
+		return err
+	}
+	metrics["PollCount"].value = strconv.FormatInt(pollCountValue+1, 10)
+
+	metrics["RandomValue"].value = strconv.FormatFloat(rand.Float64(), 'f', -1, 64)
+	mutex.Unlock()
+	return nil
+}
+
+func sendQueryUpdateMetric(client *http.Client, mName string, m metric, endpoint string) error {
+
+	builder := strings.Builder{}
+	builder.WriteString("http://")
+	builder.WriteString(endpoint)
+	builder.WriteString("/update/")
+	builder.WriteString(m.metricType)
+	builder.WriteString("/")
+	builder.WriteString(mName)
+	builder.WriteString("/")
+	builder.WriteString(m.value)
+	url := strings.TrimSpace(builder.String())
+
+	defer builder.Reset()
+
+	req, err := http.NewRequest(http.MethodPost, url, nil)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Content-Type", "text/plain")
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
+
+	return nil
+}
+
+func updateStatistic(interval int, mutex *sync.Mutex) {
+	defer wg.Done()
+	stats := &runtime.MemStats{}
+
+	ticker := time.NewTicker(time.Duration(interval) * time.Second)
+	for range ticker.C {
+
+		err := updateMetrics(Metrics, stats, mutex)
+		if err != nil {
+			panic(err)
+		}
+	}
+}
+
+func sendReport(client *http.Client, endpoint string, interval int, mutex *sync.Mutex) {
+	defer wg.Done()
+	ticker := time.NewTicker(time.Duration(interval) * time.Second)
```

Comment:
[CHANGE REQUEST] It should be `Duration`

Reply:
done

Reply:
done